src/storage-client/src/types/connections.rs

```diff
@@ -516,15 +520,25 @@ impl KafkaConnection {
         // The downside of this approach is it produces a generic error message like
         // "metadata fetch error" with no additional details. The real networking
         // error is buried in the librdkafka logs, which are not visible to users.
-        //
-        // TODO(benesch): pull out more error details from the librdkafka logs and
-        // include them in the error message.
-        mz_ore::task::spawn_blocking(
+        let (consumer, result) = mz_ore::task::spawn_blocking(
             || "kafka_get_metadata",
-            move || consumer.fetch_metadata(None, DEFAULT_FETCH_METADATA_TIMEOUT),
+            move || {
+                let result = consumer.fetch_metadata(None, DEFAULT_FETCH_METADATA_TIMEOUT);
+                (consumer, result)
+            },
         )
-        .await??;
-        Ok(())
+        .await?;
+        match result {
+            Ok(_) => Ok(()),
+            // The error returned by `fetch_metadata` does not provide any details which makes for
+            // a crappy user facing error message. For this reason we attempt to grab a better
+            // error message from the client context, which should contain any error logs emitted
+            // by librdkafka, and fallback to the generic error if there is nothing there.
+            Err(err) => match consumer.context().inner().take_err() {
```

Comment:
the only place we `take_err` is when fetching metadata...do we have plans to also take errors as part of normal consumption? this ends up being more important in sinks than sources, right?